README.md
# BSONStream

Read a binary stream that contains BSON objects and emit each as a JavaScript
object.


## Example

Write each object of a BSON file created with mongodump to stdout:

```js
var fs = require('fs');
var BSONStream = require('bson-stream');

var rs = fs.createReadStream('/some/data.bson');

rs.pipe(new BSONStream()).on('data', function(obj) {
  console.log(obj);
});
```

Example with an object containing a Buffer object as a value. The Buffer will be
embedded when serialized by the BSON npm:

```js
var assert = require('assert');
var BSON = require('bson');
var BSONStream = require('bson-stream');

var buf = new Buffer([65, 67, 70]); // 'ACF'
var obj = { foo: buf };

var bs = new BSONStream();
bs.on('data', function(obj) {
  // the original Buffer value is embedded in a new BSON Buffer object and
  // stored at the `buffer` key
  assert.strictEqual(obj.foo.buffer.toString(), 'ACF');
  console.log('ok');
});

bs.end(BSON.serialize(obj));
```


## Installation

```sh
$ npm i bson-stream
```


## API

### BSONStream([opts])
* [opts] {Object} object containing optional parameters

opts:
* raw {Boolean, default false} whether to emit JavaScript objects or raw Buffers
* maxDocLength {Number, default 16777216} maximum BSON document size in bytes
* maxBytes {Number, default infinite} maximum number of bytes to receive
* debug {Boolean, default false} whether to do extra console logging or not
* hide {Boolean, default false} whether to suppress errors or not (used in tests)

Read a binary stream that contains BSON objects and emit each as a JavaScript
object (or as a Buffer if opts.raw is true).

Note: implements the [BSON specification], as supported by [js-bson].
Note2: Buffer objects will be embedded in a new BSON Buffer object at `buffer`.


## Tests

```sh
$ npm test
```


## License

ISC

Copyright (c) 2014, 2015 Tim Kuijsten

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted, provided that the above
copyright notice and this permission notice appear in all copies.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.

[BSON specification]: http://bsonspec.org/spec.html
[js-bson]: https://www.npmjs.org/package/bson


ChangeLog
2020.10.23, Version 1.0.0

* update bson dependency from ^0.4.20 to ^4.2.0


2015.12.29, Version 0.0.8

* fix use of browserify with bson npm module


2015.12.24, Version 0.0.7

* make compatible with browserify


2015.08.31, Version 0.0.6

* fix compatibility with bson 0.x (namely 0.4.x)
* Readme: markup


2015.01.09, Version 0.0.5

* update license from MIT to ISC
* document and test Buffer values in objects


2014.11.18, Version 0.0.4

* test opts.maxBytes type checking
* readme: simpler example


2014.11.14, Version 0.0.3

* support maxBytes option


2014.10.27, Version 0.0.2

* clarify README
* package.json: fix links to repo


2014.10.27, Version 0.0.1

* public release


test/index.js
/**
 * Copyright (c) 2014, 2015 Tim Kuijsten
 *
 * Permission to use, copy, modify, and/or distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

'use strict';

/*jshint -W068 */

var fs = require('fs');
var crypto = require('crypto');

var should = require('should');
var BSON = require('bson');

var BSONStream = require('../index.js');

describe('BSONStream', function() {
  it('should require opts to be an object', function() {
    (function() { var bs = new BSONStream(''); return bs; }).should.throw('opts must be an object');
  });

  it('should require opts.raw to be a boolean', function() {
    (function() { var bs = new BSONStream({ raw: '' }); return bs; }).should.throw('opts.raw must be a boolean');
  });

  it('should require opts.maxDocLength to be a number', function() {
    (function() { var bs = new BSONStream({ maxDocLength: '' }); return bs; }).should.throw('opts.maxDocLength must be a number');
  });

  it('should require opts.maxBytes to be a number', function() {
    (function() { var bs = new BSONStream({ maxBytes: '' }); return bs; }).should.throw('opts.maxBytes must be a number');
  });

  it('should require opts.debug to be a boolean', function() {
    (function() { var bs = new BSONStream({ debug: '' }); return bs; }).should.throw('opts.debug must be a boolean');
  });

  it('should require opts.hide to be a boolean', function() {
    (function() { var bs = new BSONStream({ hide: '' }); return bs; }).should.throw('opts.hide must be a boolean');
  });

  it('should construct', function() {
    var bs = new BSONStream();
    return bs;
  });

  it('should be a writable stream', function(done) {
    var bs = new BSONStream();
    bs.end(done);
  });

  it('should be a readable stream', function(done) {
    var bs = new BSONStream();
    bs.resume();
    bs.on('end', done);
    bs.end();
  });

  it('should emit one valid empty BSON buffer', function(done) {
    var bs = new BSONStream({ raw: true });
    bs.on('data', function(data) {
      should.strictEqual(data instanceof Buffer, true);
      should.strictEqual(data.length, 5);
      should.strictEqual(data[0], 0x05);
      should.strictEqual(data[1], 0x00);
      should.strictEqual(data[2], 0x00);
      should.strictEqual(data[3], 0x00);
      should.strictEqual(data[4], 0x00);
      should.strictEqual(data[5], undefined);
      done();
    });
    bs.end(new Buffer([0x05, 0x00, 0x00, 0x00, 0x00]));
  });

  it('should emit one valid empty object', function(done) {
    var bs = new BSONStream();
    bs.on('data', function(obj) {
      should.deepEqual(obj, {});
      done();
    });
    bs.end(new Buffer([0x05, 0x00, 0x00, 0x00, 0x00]));
  });

  it('should err when more than maxBytes are written', function(done) {
    var bs = new BSONStream({ raw: true, maxBytes: 4 });
    bs.on('error', function(err) {
      should.strictEqual(err.message, 'more than maxBytes received');
      done();
    });
    bs.on('data', function() { throw Error('incomplete BSON emitted'); });
    bs.end(new Buffer([0x05, 0x00, 0x00, 0x00, 0x00]));
  });

  it('should require a document length of at least 5 bytes', function(done) {
    var bs = new BSONStream({ maxBytes: 5 });
    bs.on('error', function(err) {
      should.strictEqual(err.message, 'invalid document length');
      done();
    });
    bs.on('data', function() { throw Error('incomplete BSON emitted'); });
    bs.end(new Buffer([0x04, 0x00, 0x00, 0x00, 0x00]));
  });

  it('should require each document to not exceed the default maximum of 16 MB', function(done) {
    var bs = new BSONStream();
    bs.on('data', function() { throw Error('incomplete BSON emitted'); });
    bs.on('error', function(err) {
      should.strictEqual(err.message, 'document exceeds configured maximum length');
      done();
    });
    bs.end(new Buffer([0x01, 0x00, 0x00, 0x01]));
  });

  it('should require each document to not exceed the configured maximum', function(done) {
    var bs = new BSONStream({ maxDocLength: 5 });
    bs.on('data', function() { throw Error('incomplete BSON emitted'); });
    bs.on('error', function(err) {
      should.strictEqual(err.message, 'document exceeds configured maximum length');
      done();
    });
    bs.end(new Buffer([0x06, 0x00, 0x00, 0x00]));
  });

  it('should require that a document terminates with 0x00', function(done) {
    var bs = new BSONStream();
    bs.on('error', function(err) {
      should.strictEqual(err.message, 'invalid document termination');
      done();
    });
    bs.on('data', function() { throw Error('incomplete BSON emitted'); });
    bs.end(new Buffer([0x05, 0x00, 0x00, 0x00, 0x01]));
  });

  it('should not emit documents when document length is not reached', function(done) {
    var bs = new BSONStream();
    bs.on('data', function() { throw Error('incomplete BSON emitted'); });
    bs.on('end', done);
    bs.end(new Buffer([0x00, 0x00, 0x00, 0x01, 0x00]));
  });

  it('should deserialize a generated BSON buffer correctly', function(done) {
    var obj = {
      foo: 'bar',
      bar: 42,
      baz: false,
      qux: null
    };

    var bs = new BSONStream();
    bs.on('data', function(data) {
      should.deepEqual(data, {
        foo: 'bar',
        bar: 42,
        baz: false,
        qux: null
      });
      done();
    });
    bs.end(BSON.serialize(obj));
  });

  it('should deserialize two generated BSON buffer correctly', function(done) {
    var obj1 = {
      foo: 'bar'
    };

    var obj2 = {
      foo: 'baz',
      bar: 42,
      baz: false,
      qux: null
    };

    var bs = new BSONStream();

    var arr = [];

    bs.on('data', function(data) {
      arr.push(data);
    });

    bs.on('end', function() {
      should.strictEqual(arr.length, 2);
      should.deepEqual(arr[0], {
        foo: 'bar'
      });
      should.deepEqual(arr[1], {
        foo: 'baz',
        bar: 42,
        baz: false,
        qux: null
      });
      done();
    });

    bs.end(Buffer.concat([BSON.serialize(obj1), BSON.serialize(obj2)]));
  });

  it('should skip noise in previous chunks and emit two generated BSON buffers', function(done) {
    // this only works because the "noise" indicates a document length that exceeds the default allowed maximum length of 16 MB
    var noise = new Buffer([9,4,5,8,4,6,2,5,4]);

    var obj1 = {
      foo: 'bar'
    };

    var obj2 = {
      foo: 'baz',
      bar: 42,
      baz: false,
      qux: null
    };

    var bs = new BSONStream();

    var arr = [];

    bs.on('data', function(data) {
      arr.push(data);
    });

    bs.on('error', function(err) {
      should.strictEqual(err.message, 'document exceeds configured maximum length');
    });

    bs.on('end', function() {
      should.strictEqual(arr.length, 2);
      should.deepEqual(arr[0], {
        foo: 'bar'
      });
      should.deepEqual(arr[1], {
        foo: 'baz',
        bar: 42,
        baz: false,
        qux: null
      });
      done();
    });

    bs.write(Buffer.concat([noise]));
    bs.write(Buffer.concat([BSON.serialize(obj1), noise, BSON.serialize(obj2)]));
    bs.write(Buffer.concat([BSON.serialize(obj2)]));
    bs.end();
  });

  it('stream binary data from disk', function(done) {
    fs.readFile(__dirname + '/image.png', function(err, data) {
      if (err) { throw err; }

      should.strictEqual(data.length, 121271);

      var shasum = crypto.createHash('sha1');
      shasum.update(data);
      should.strictEqual(shasum.digest('hex'), '6b7e9149ba026c6b413f40f5d4ac02fd57cdd57c');

      var obj = { foo: data };

      var bs = new BSONStream();

      var arr = [];

      bs.on('data', function(data) {
        arr.push(data);
      });

      bs.on('error', function(err) {
        should.strictEqual(err.message, 'document exceeds configured maximum length');
      });

      bs.on('end', function() {
        should.strictEqual(arr.length, 1);
        should.strictEqual(arr[0].foo.buffer.length, 121271);

        var shasum2 = crypto.createHash('sha1');
        shasum2.update(arr[0].foo.buffer);
        should.strictEqual(shasum2.digest('hex'), '6b7e9149ba026c6b413f40f5d4ac02fd57cdd57c');
        done();
      });

      bs.end(BSON.serialize(obj));
    });
  });
});


index.js
/**
 * Copyright (c) 2014, 2015 Tim Kuijsten
 *
 * Permission to use, copy, modify, and/or distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

'use strict';

var util = require('util');
var Transform = require('stream').Transform;

var BSON = require('bson');

/**
* BSONStream
*
* Read a binary stream that contains BSON objects and emit each as a JavaScript
* object (or as a Buffer if opts.raw is true).
*
* Note: implements BSON specification 1.0 (http://bsonspec.org/spec.html), as
* supported by js-bson (https://www.npmjs.org/package/bson).
*
* Note2: Buffer objects will be embedded in a new BSON Buffer object at `buffer`.
*
* @param {Object} [opts] object containing optional parameters
*
* opts:
* raw {Boolean, default false} whether to emit JavaScript objects or raw Buffers
* maxDocLength {Number, default 16777216} maximum BSON document size in bytes
* maxBytes {Number, default infinite} maximum number of bytes to receive
* debug {Boolean, default false} whether to do extra console logging or not
* hide {Boolean, default false} whether to suppress errors or not (used in tests)
*/
function BSONStream(opts) {
  if (typeof opts !== 'undefined' && typeof opts !== 'object') { throw new TypeError('opts must be an object'); }
  opts = opts || {};

  if (typeof opts.raw !== 'undefined' && typeof opts.raw !== 'boolean') { throw new TypeError('opts.raw must be a boolean'); }
  if (typeof opts.maxDocLength !== 'undefined' && typeof opts.maxDocLength !== 'number') { throw new TypeError('opts.maxDocLength must be a number'); }
  if (typeof opts.maxBytes !== 'undefined' && typeof opts.maxBytes !== 'number') { throw new TypeError('opts.maxBytes must be a number'); }
  if (typeof opts.debug !== 'undefined' && typeof opts.debug !== 'boolean') { throw new TypeError('opts.debug must be a boolean'); }
  if (typeof opts.hide !== 'undefined' && typeof opts.hide !== 'boolean') { throw new TypeError('opts.hide must be a boolean'); }

  Transform.call(this, opts);

  this._maxDocLength = opts.maxDocLength || 16777216;
  // bson spec: signed int32
  if (this._maxDocLength > 2147483647) { throw new Error('maxDocLength can not exceed 2147483647 bytes'); }

  this._maxBytes = opts.maxBytes;

  this._raw = opts.raw;
  this._debug = opts.debug || false;
  this._hide = !!opts.hide;

  this._writableState.objectMode = false;
  if (!this._raw) {
    this._readableState.objectMode = true;
  }

  // initialize internal buffer
  this._reset();
}
util.inherits(BSONStream, Transform);

module.exports = BSONStream;

// reset internal buffer
BSONStream.prototype._reset = function _reset() {
  if (this._debug) { console.log('_reset'); }

  this._buffer = new Buffer(0);
  this._doclen = null;
};

// read up to doclen bytes
BSONStream.prototype._parseDocs = function _parseDocs(cb) {
  if (this._debug) { console.log('_parseDocs'); }

  // first make sure the expected document length is known
  if (!this._doclen) {
    if (this._buffer.length < 4) {
      // wait for more chunks
      cb();
      return;
    }

    // bson spec defines signed int32
    var doclen = this._buffer.readInt32LE(0);

    if (this._debug) { console.log('_parseDocs doc length', doclen); }

    // should have at least have 5 bytes since the minimum document contains an int32 and a \x00 terminal
    if (doclen < 5) {
      // discard buffer
      this._reset();
      cb(new Error('invalid document length'));
      return;
    }

    if (doclen > this._maxDocLength) {
      // discard buffer
      this._reset();
      cb(new Error('document exceeds configured maximum length'));
      return;
    }

    this._doclen = doclen;
  }

  // since the expected document length is known, make sure the complete length is in the internal buffer
  if (this._buffer.length < this._doclen) {
    // wait for more chunks
    cb();
    return;
  }

  // since the complete document is in the buffer, try to read and parse it as BSON

  // check if document ends with 0x00
  if (this._buffer[this._doclen - 1] !== 0x00) {
    // discard buffer
    this._reset();
    cb(new Error('invalid document termination'));
    return;
  }

  var rawdoc = this._buffer.slice(0, this._doclen);
  var obj;

  try {
    obj = BSON.deserialize(rawdoc);
  } catch (err) {
    // discard buffer
    this._reset();
    cb(err);
    return;
  }

  // shift document from internal buffer and nullify expected document length
  this._buffer = this._buffer.slice(this._doclen);
  this._doclen = null;

  // push the raw or parsed doc out to the reader
  this.push(this._raw ? rawdoc : obj);

  // check if there might be any new document that can be parsed
  if (this._buffer.length > 4) {
    this._parseDocs(cb);
  } else {
    cb();
  }
};

BSONStream.prototype._transform = function _transform(chunk, encoding, cb) {
  if (this._debug) { console.log('_transform', chunk); }

  var newLength = this._buffer.length + chunk.length;

  if (this._maxBytes && newLength > this._maxBytes) {
    cb(new Error('more than maxBytes received'));
    return;
  }

  this._buffer = Buffer.concat([this._buffer, chunk], newLength);
  this._parseDocs(cb);
};
